tests/lac/tensor_product_matrix_07.cc

```diff
@@ -0,0 +1,207 @@
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+// Test TensorProductMatrixSymmetricSum for zero (constrained) rows and colums.
+
+#include <deal.II/dofs/dof_handler.h>
+
+#include <deal.II/fe/fe_dgq.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/fe/mapping_q1.h>
+
+#include <deal.II/grid/grid_generator.h>
+
+#include <deal.II/lac/full_matrix.h>
+#include <deal.II/lac/tensor_product_matrix.h>
+#include <deal.II/lac/vector.h>
+
+#include "../tests.h"
+
+#include "../testmatrix.h"
+
+template <int dim, typename Number>
+std::tuple<FullMatrix<Number>, FullMatrix<Number>>
+compute_reference_matrices(const unsigned int fe_degree)
+{
+  MappingQ1<dim> mapping;
+  QGauss<dim>    quadrature(fe_degree + 1);
+  FE_DGQ<dim>    fe(fe_degree);
+
+  Triangulation<dim> tria;
+  GridGenerator::hyper_cube(tria);
+
+  DoFHandler<dim> dof_handler(tria);
+  dof_handler.distribute_dofs(fe);
+
+  const unsigned int n_dofs_1D = fe.n_dofs_per_cell();
+
+  FullMatrix<Number> mass_matrix_reference(n_dofs_1D, n_dofs_1D);
+  FullMatrix<Number> derivative_matrix_reference(n_dofs_1D, n_dofs_1D);
+
+  FEValues<dim> fe_values(mapping,
+                          fe,
+                          quadrature,
+                          update_values | update_gradients | update_JxW_values);
+
+  fe_values.reinit(dof_handler.begin());
+
+  for (const unsigned int q_index : fe_values.quadrature_point_indices())
+    for (const unsigned int i : fe_values.dof_indices())
+      for (const unsigned int j : fe_values.dof_indices())
+        {
+          mass_matrix_reference(i, j) +=
+            (fe_values.shape_value(i, q_index) *
+             fe_values.shape_value(j, q_index) * fe_values.JxW(q_index));
+
+          derivative_matrix_reference(i, j) +=
+            (fe_values.shape_grad(i, q_index) *
+             fe_values.shape_grad(j, q_index) * fe_values.JxW(q_index));
+        }
+
+
+
+  return {mass_matrix_reference, derivative_matrix_reference};
+}
+
+template <typename Number>
+void
+print(const FullMatrix<Number> &matrix)
+{
+  matrix.print_formatted(deallog.get_file_stream(), 2, false, 7);
+  deallog << std::endl << std::endl;
+}
+
+
+void
+do_test(const bool zero_out_constraints)
+{
+  using Number                 = double;
+  const unsigned int dim       = 2;
+  const unsigned int n_rows_1d = 4;
+
+  // compute 2D stiffness matrix
+  const auto refernce_matrices_2D =
+    compute_reference_matrices<dim, Number>(n_rows_1d - 1);
+  auto K_2D = std::get<1>(refernce_matrices_2D);
+
+  // ... and apply DBC
+  for (unsigned int j = 0; j < Utilities::pow(n_rows_1d, dim); ++j)
+    for (unsigned int i = 0; i < Utilities::pow(n_rows_1d, dim - 1); ++i)
```

Comment:
I believe the reason to set two different loop lengths here is that you pick boundary conditions on a single surface in the lexicographic order (x-direction), and leave the rest unchanged. This is simple to miss, so I would like to have a more extensive comment or an descriptive variable name of the loop length referring the unknowns at the identified boundary.

Reply:
I have added a comment.